reana_server/complexity.py

```diff
@@ -0,0 +1,360 @@
+                complexity = stage_value["complexity"]
+
+                # Handle nested stages
+                parsed_stages = _parse_stages(stage_value["stages"])
+                stage_value["stages"] = parsed_stages
+                if parsed_stages:
+                    complexity = self._calculate_complexity(parsed_stages)
+
+                # Handle scatter parameters
+                if stage_value["scatter_params"]:
+                    first_param = stage_value["scatter_params"][0]
+                    param_len = len(stage_value["params"].get(first_param, []))
+                    complexity = [(item[0] * param_len, item[1]) for item in complexity]
+
+                stage_value["complexity"] = complexity
+            return stages
+
+        return _parse_stages(stages)
+
+    def parse_specification(self, initial_step):
+        """Parse Yadage workflow specification tree."""
+        steps = self._parse_steps(self.specification["stages"], initial_step)
+        steps = self._populate_parameters(steps, self.input_params)
+        steps = self._populate_complexity(steps)
+        return steps
+
+
+class CWLComplexityEstimator(ComplexityEstimatorBase):
+    """REANA CWL workflow complexity estimation."""
+
+    def _parse_steps(self, workflow):
+        """Parse CWL workflow specification tree."""
+        tree = {}
+        steps = workflow.get("steps", [])
+        for step in steps:
+            name = step.get("id")
+            hints = step.get("hints", [{}]).pop()
+            # Parse scatter params
+            scatter = step.get("scatter")
+            scatter_params = None
+            if scatter:
+                scatter_params = next(
+                    filter(lambda p: p["id"] == scatter, step.get("in", [])), {},
+                ).get("source")
+            # Parse nested workflows
+            workflow_file = step.get("run")
+            nested_workflow = workflow_file if isinstance(workflow_file, str) else None
+            # Parse initial complexity
+            jobs = self._get_number_of_jobs(hints)
+            memory_limit = self._get_memory_limit(hints)
+            complexity = [(jobs, memory_limit)]
+            # Parse params and dependencies
+            params = list(map(lambda i: i.get("source", ""), step.get("in", [])))
+            dependencies = set()
+            for param in params:
+                # Extract dependencies from param (e.g '#main/gendata/data')
+                if param:
+                    dependencies.update(param.split("/")[1:-1])
+
+            tree[name] = {
+                "complexity": complexity,
+                "params": params,
+                "dependencies": list(dependencies),
+                "scatter_params": scatter_params,
+                "workflow": nested_workflow,
+            }
+        return tree
+
+    def _parse_workflow(self, workflow):
+        """Parse CWL workflow specification tree."""
+        tree = {}
+        if isinstance(workflow, dict):
```

Comment:
I highly recommend using our autogenerated classes at https://github.com/common-workflow-language/cwl-utils/blob/main/cwl_utils/parser_v1_1.py instead of parsing CWL yourself. I can see that this code doesn't support CWL sub-workflows, for example.

Reply:
Thanks a lot for your suggestion @mr-c ! We will check the parser out and hopefully integrate it into our code. Funny enough, I am implementing nested workflow support at the moment :)

Reply:
@mr-c, what would be the main benefit of using this parser from `cwl_utils`? Currently we are parsing the `$graph` which we get by calling `cwltool --pack --quiet workflow_file`. I just checked the output of `cwl_utils` parser and it gives pretty much the same information as the one which we use. Maybe I'm missing something?

Reply:
In general, the benefit is concrete classes, and that the schema-salad transformations are already performed (type DSLs expanded to their normalized form, ` $include` and `$import` acted upon). Likewise the document undergoes basic validation and any errors result in specific messages. However, you do get almost all of that from `cwltool --pack` besides the nice concrete classes instead of untyped dictionaries. So maybe there isn't enough to justify the change in this instance.

However, if later other types of analysis is needed, then you may benefit from the `cwl-utils` approach, so maybe leave a comment mentioning that?

Reply:
Thanks for the explanation! I guess we can revive this topic once upgrading the [cwltool version to the latest](https://github.com/reanahub/reana-workflow-engine-cwl/issues/187)